### ISSUE TITLE
GPTJ output not consistent with that of transformers

### ISSUE THREAD

**Issue by alexandred** [vllm-project/vllm#590]
Hey guys,

I've been trying to get the base GPT-J model (EleutherAI/gpt-j-6b) served using vLLM, but I am running into issues with the quality of the output. Using the same offline inference code as in https://github.com/vllm-project/vllm/blob/main/examples/offline_inference.py, here is what I obtain from the model (running on an RTX A5000):

``` python3
from vllm import LLM, SamplingParams

sampling_params = SamplingParams(temperature=0.8, top_p=0.95, max_tokens=128)
llm = LLM(model="EleutherAI/gpt-j-6b")
prompts = ["AI is going to"] * 4
outputs = [t.outputs[0].text for t in llm.generate(prompts, sampling_params)]

output = [" get a lot of new stuff. A lot of other new stuff\nthat's been added and things are going to come out.\nthat you know, it's going to be implemented, that's coming out.\nin the\nand we've got like a lot of stuff coming out, and things in the like AI stuff. So we're going to be doing.\nbefore the new stuff\nand some of the AI, the future.\nthe AI\nis added in the next year and we're going to the next year.\nthat's going to be in the next gen.\nand it's going to be added",
 " change the way we work, not just the way we do business, but also how we work and we live. How we work, the way we live. In the way we do our everyday activities. It will be. We think, what we live.\n\n\n\n\n\n\n\nthe way we interact with each other, but our living.\n\nthe way we relate to each other, but how we live. AI.\n\n\nit's not even our lives. The economy, live, and build our daily life.\nhow we play, but also our lives and engage with each other, and how we",
 " be our friend.\n\n\n\n\nHere's a good reason we're gonna do this\nAnd you know, and that was interesting.\nWe'll have to be, like,\nour employees are gonna start.\nOur\nThey did you know, as you know.\nTotally, we're gonna have like you know\nAnd that's the oil.\nSo that some other question, like, and that we're gonna\nuse and and I'm going to be in the most of the middle of\nThe boss and we're going to be on TV,\nTogether. I'm gonna be\nWe're gonna",
 " save us all. Some of them, anyway. It will. AI is almost certainly. It's just a utopianism is what we can see how much about as serious, but it will be a way out of the future. I mean.\n\n\nsmart, it's going to out how it will be their business is the father than we're going to the ultimate evolution. I think we're going to be the in a perfect, the machine.\n\nnot be a bit more than 3Dot-it's just it.\n\n\n\nkill us we have long as we're going to have it will be making"]
 ```

For comparison's sake, here is what I obtain using transformers:

```python3
from transformers import AutoModelForCausalLM, AutoTokenizer
model = AutoModelForCausalLM.from_pretrained("EleutherAI/gpt-j-6b").half().eval().to(0)
tokenizer = AutoTokenizer.from_pretrained("EleutherAI/gpt-j-6b")
inputs = tokenizer("AI is going to", return_tensors="pt").to(0)
params = {"temperature": 0.85, "top_p": 0.94, "max_new_tokens": 128, "do_sample": True}
outputs = []
for _ in range(1,4):
    out = model.generate(inputs["input_ids"], **params)
    outputs.append(tokenizer.decode(out[0]))
outputs = [
'AI is going to be our future and this is going to be a big part of that," said Peter Lee, vice president of Google\'s self-driving cars and mapping division.\n\n"It will allow us to make the most efficient use of public roads that are more congested and have more traffic," Lee told the BBC. "It will allow us to better respond to traffic conditions and accidents."\n\nIt\'s certainly a big deal, and one that Google will have to work its way through. The company is a pioneer in the area of driverless cars, but its software has been unable to operate as a taxi driver.\n\nIts self',
'AI is going to make a comeback. It's already making a comeback. The best technology to understand AI is not the technology itself. It's human understanding of what the technology is. We're just starting to understand how human beings think, and the way that human beings think is going to revolutionize the way we think.\n\nIt's hard for us to imagine the future that AI brings. We're seeing AI in every form, and we're seeing it becoming a reality in our lives. You may not see it right away, but it's right around the corner.\n\n',
'AI is going to be a game changer and a very important factor in the success of the entire Internet. It will not be just another technology, but a game changing technology. The way we do things has been changed.\n\nAnd we can't wait to help you succeed in the online world!\n\nOur Vision\n\nOur vision is to provide the best eCommerce platform that can help your business grow. We believe in the concept of "Customer First". We will work with you and your team to deliver the best online experience.\n\nWe will continue to innovate and develop new features, functionality and services to help you',
'AI is going to change everything. The world of medicine and the world of business will be forever changed.\n\nThe world of medicine will change drastically. Doctors will be able to diagnose illnesses and provide immediate treatment without needing to send their patients to specialists. A patient's condition will be known almost instantly and he/she will be able to receive the best possible treatment.\n\nThe world of business will change drastically as well. With the help of AI, companies can easily increase their efficiency, improve their products, and ultimately increase profits.\n\nWhy is AI so important?\n\nAI has become incredibly important because it can help us solve many'
]
```

I've noticed this behaviour independent of the choice of sampling parameters.

Any help in resloving this is greatly appreciated!

**Comment by zhuohan123** [comment]
Hi! Can you try some deterministic sampling methods like greedy decoding and compare the results? To be honest I cannot easily tell the differences between the qualities of two sampling results.

**Comment by alexandred** [comment]
Thanks for getting back to me!

@AlekseyKorshuk and I ran the inference with deterministic sampling and got the following results:

```python3
sampling_params = SamplingParams(temperature=0, top_k=-1, max_tokens=128)
llm = LLM(model="philschmid/gpt-j-6B-fp16-sharded")
prompts = ["Obama was born in [Honolulu](https://en.wikipedia.org/wiki/Honolulu), Hawaii. After graduating from [Columbia University](https://en.wikipedia.org/wiki/Columbia_University) in 1983, he worked as a [community organizer](https://en.wikipedia.org/wiki/Community_organizing) in Chicago. In 1988, he enrolled in [Harvard Law School](https://en.wikipedia.org/wiki/Harvard_Law_School), where he was the first black president of the [Harvard Law Review](https://en.wikipedia.org/wiki/Harvard_Law_Review). After graduating,"] * 4
outputs = [t.outputs[0].text for t in llm.generate(prompts, sampling_params)]
outputs = [
' he was a year later, he was a year. He was a student of the University of Chicago, he was a community organizer for the University of Chicago, he was a community organizer, he was a year, and then he was a year later, he was a year.\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n',
 ' he was a year later, he was a year. He was a student of the University of Chicago, he was a community organizer for the University of Chicago, he was a community organizer, he was a year, and then he was a year later, he was a year.\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n',
 ' he was a year later, he was a year. He was a student of the University of Chicago, he was a community organizer for the University of Chicago, he was a community organizer, he was a year, and then he was a year later, he was a year.\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n',
 ' he was a year later, he was a year. He was a student of the University of Chicago, he was a community organizer for the University of Chicago, he was a community organizer, he was a year, and then he was a year later, he was a year.\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n']
```

Using the same prompt and model on HG (using do_sample=False), we got the following completion:
`Obama served as an attorney for the Illinois State Senate and then taught constitutional law at the University of Chicago.\n`

**Comment by leegohi04517** [comment]
For us, VLLM has greatly improved in terms of inference speed, but the difference in the inference results due to the different sampling methods between VLLM and Transformer is quite confusing. Transformer has a good ecosystem and there is consensus on the sampling parameters. Can it support the option of choosing between VLLM sampling and Transformer sampling methods?

**Comment by syskn** [comment]
VLLM's outputs are definitely (significantly) more repetitive than HF inference (as you can see in Alex's results in the first post and the greedy decoding result). I think sampling process itself has no significant delta; it might be something in cuda kernels or attention.

Related topic I posted in discussions:
https://github.com/vllm-project/vllm/discussions/471

**Comment by ri938** [comment]
yeah it looks to me like GPTJ responses are poor quality.

I worry that its not just an issue with the GPTJ implementation but is a more general regression in quality in vLLM.

**Comment by juliensalinas** [comment]
Same problem here with GPT-J.
I am using the default sampling params from vLLM. I also tried playing with all the params availble, but it did not help.

I am consistently getting semi-gibberish outputs.
Here is one example.

Input:
> Let me tell you a story:

Output:
> Let me tell you a story:\n\nIt’s midnight. Your eyes begin to fall right at noon, I came out of Scotland a summer, right at the dock where this story take in the sky blue.\n\nAM,\nNever felt feverich I never...

**Comment by PanQiWei** [comment]
May be it's because GPT-J's tensor rotate logic is different to other models, as described in #747

**Comment by chiragjn** [comment]
+1
Noticing similar weird issues, with both GPT-J 6B and Pygmalion 6B (finetuned on top of GPT-J)
```python
import torch
from transformers import pipeline

MODEL_ID = "EleutherAI/gpt-j-6b"
PROMPT = """
import tensorflow
# 4 layer CNN with a softmax output
# test on MNIST data set
"""
prompts = [PROMPT]
p = pipeline(
    "text-generation", model=MODEL_ID, tokenizer=MODEL_ID, torch_dtype=torch.float16, device=1
)
out = p(
    prompts, temperature=0.8,  top_p=0.95,  max_new_tokens=100, top_k=10000
)
print(out[0][0]["generated_text"])
```

which produces a reasonable output
```
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.

import tensorflow
# 4 layer CNN with a softmax output
# test on MNIST data set

import numpy as np
import matplotlib.pyplot as plt

from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split

from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation
from keras.layers import Conv2D, MaxPooling2D, Flatten
from keras.optimizers import SG
```

But with vLLM, something is off

```python
from vllm import LLM, SamplingParams
llm = LLM(model=MODEL_ID)
sampling_params = SamplingParams(temperature=0.8, top_p=0.95, max_tokens=100, top_k=10000)
outputs = llm.generate(prompts, sampling_params)
for output in outputs:
    prompt = output.prompt
    generated_text = output.outputs[0].text
    print(f"Prompt: {prompt!r}, Generated text: {generated_text!r}")
print(generated_text)
```

```
import tensorflow
from tensorflow as TF-layers
from tensorflow_model
import tensorflow
import time
import nn
import tensorflow as tf.nn.data

from tensorflow as tf
from tensorflow
import tensorflow as tf
from tensorflow as nn

# from Tensorflow
from tensorflow as tf.image classification
import tf
import tensorflow as tf
```

**Comment by WoosukKwon** [comment]
Hi @alexandred @chiragjn @juliensalinas @ri938 , thanks for letting us know the bug. This is indeed due to my misunderstanding of the GPT-J RoPE, which is different from GPT-NeoX's RoPE. The bug is fixed by #941. I've checked that after the bug fix, the generated outputs by vLLM match the HF's outputs when using argmax sampling. Apologies for the confusion and inconvenience.
